DAT-455 Un-delete via button

madelinekosse opened the pull request:
When a dataset is deleted, an admin can now re-instate it via a button on the top of the 'edit' page like this:

![image](https://github.com/GreaterLondonAuthority/dfl-ckanext/assets/20282373/25bd3e54-451b-4fa9-985a-ba010a5bfdbe)

When clicked, the dataset is un-deleted and the user is redirected to the dataset view page

madelinekosse committed:
DAT-455 Un-delete via button

When a dataset is deleted, an admin can now re-instate it via a button on the
top of the 'edit' page.

kjoshi reviewed (commented):
Thanks

This mostly looks good to me, but I'm sometimes getting a confusing error about the `data_quality` field being missing when I try and undelete a dataset.

If I do it the old way by setting the dataset state to `active` and clicking `Update`, everything works Ok.

https://github.com/GreaterLondonAuthority/dfl-ckanext/assets/1217533/d4ad8622-6824-4d66-a34a-493517721c98

But if I try and use the `Undelete` button I get an error, even though the `Data Quality` value is the same both times.

https://github.com/GreaterLondonAuthority/dfl-ckanext/assets/1217533/363ba677-ec49-4376-9b75-bada92a083a5

Did you see this when you were testing?

madelinekosse commented:
> I'm sometimes getting a confusing error about the `data_quality` field being missing when I try and undelete a dataset.
> 
> If I do it the old way by setting the dataset state to `active` and clicking `Update`, everything works Ok.
> 
> But if I try and use the `Undelete` button I get an error, even though the `Data Quality` value is the same both times.
> Did you see this when you were testing?

Thanks for pointing that out - I have seen it when doing other things such as adding/removing a resource and thought it was because the field was required, but now I see that `None` is a valid value and. I know I could fix the issue for this PR by passing the data quality field through to the endpoint when you click the button, but that doesn't sound quite right as we'll probably keep running into it as we add more admin action stuff. Will investigate why it's happening

kjoshi commented:
Yeah I don't think you need to and anything to your `package_patch` call either. The docs say that `package_patch` should only update the fields that have been provided. I don't understand why it's trying to re-validate the whole form. Or why the validation is doing something different than what you get when you do `Update dataset` 🤔

madelinekosse commented:
> Yeah I don't think you need to and anything to your `package_patch` call either. The docs say that `package_patch` should only update the fields that have been provided. I don't understand why it's trying to re-validate the whole form. Or why the validation is doing something different than what you get when you do `Update dataset` 🤔

It seems like we can't add a dataset without the data_quality field at all; I changed the validator to allow it but there's something else going on because the field is still marked as missing

madelinekosse committed:
Fix "data_quality: missing_value" error

When using the undelete button on a dataset that has a `data_quality` value of
`None`, we were seeing a missing value error; this was also happening when
adding a dataset or resource despite the schema permitting `None` as a value for
the custom field.

A perplexing issue because using `dataset_update` to edit a dataset with the
exact same values worked just fine - unfortunately I have not yet figured out
why this was happening, but have fixed it by changing the validators for the
field in question.

The problem seems to have been with the `convert_to_extras` converter, which
cannot handle empty fields. As we see in the [docstring for the ignore empty
converter](https://github.com/ckan/ckan/blob/60faa6b5061b3c9dd3ef6971245ef8a5e4d14c16/ckan/lib/navl/validators.py#L235),
CKAN needs the `ignore_empty` or `ignore_missing` validator to appear __before__
the other validators or converters and works by omitting other
validation/conversion steps appearing after it. It's no good having a
validator that can return `None` because the `convert_to_extras` can't handle
this value.

An additional note, because it's not obvious: the `ignore_missing` validator did
not work in this case, and won't work for any form input field, because the
value passed from the form is the empty string, not `None`. It's `ignore_empty`
that we should be using for this and all future optional form fields.

kjoshi commented:
Works perfectly now!
Thanks for the details too. Weird that it was working for `package_update` 🤔 
But what we've got now seems more correct than what we had before 👍
